keps/sig-scheduling/785-scheduler-component-config-api/README.md

```diff
@@ -0,0 +1,110 @@
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [x] Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [x] KEP approvers have approved the KEP status as `implementable`
+- [x] Design details are appropriately documented
+- [x] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [x] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The kube-scheduler configuration API `kubescheduler.config.k8s.io` is currently
+in version `v1alpha2`. We propose its graduation to `v1beta1` in order to
+promote its wider use.
+
+## Motivation
+
+The `kubescheduler.config.k8s.io` API has been in alpha stage for several
+releases. In release 1.18, we introduced `v1alpha2`, including important
+changes such as:
+
+- The removal of the old Policy API in favor of plugin configurations, that
+  align with the new scheduler framework.
+- The introduction of scheduling profiles, that allow a scheduler to appear
+  as multiple schedulers under different configurations.
+  
+A configuration API allows cluster administrators to build, validate and
+version their configurations in a more robust way than using command line flags.
+
+Graduating this API to Beta is a sign of its maturity that would encourage wider
+usage.
+
+### Goals
+
+- Introduce `kubescheduler.config.k8s.io/v1beta1` as a copy of
+`kubescheduler.config.k8s.io/v1alpha2` with minimal cleanup changes.
+- Use the newly created API objects to build the default configuration for kube-scheduler.
+
+### Non-Goals
+
+- Remove support for `kubescheduler.config.k8s.io/v1alpha2`
+
+## Proposal
+
+For the most part, `kubescheduler.config.k8s.io/v1beta1` will be a copy of
+`kubescheduler.config.k8s.io/v1alpha2`, with the following differences:
+
+- [ ] `.bindTimeoutSeconds` will be an argument for `VolumeBinding` plugin.
+- [ ] `.profiles[*].plugins.unreserve` will be removed. We will preserve
+  the extension in the internal type as long as `v1alpha2` is not removed.
+  During conversion of `v1beta1`, all the plugins from `reserve` will be
+  copied into `unreserve`.
+- [ ] Embedded types of `RequestedToCapacityRatio` will include missing json tags
+  and will be decoded with a case-sensitive decoder.
+
+### Risks and Mitigations
+
+The major risk is around the removal of the `unreserve` extension point.
```

Comment:
Can you clarify what an out-of-tree plugin that wanted to just register for reserve or unreserve would have to do to work properly in a v1beta1 config? Is the plugin invoked with enough information to be able to tell whether the call is a reserve or unreserve call and whether it should ignore the one it doesn't care about?

Reply:
The plugin is invoked with different function calls for Reserve and Unreserve.

When using v1alpha2, Reserve and Unreserve can be enabled independently. When using v1beta1, they are enabled together. The plugins don't need to know the version. When using v1beta1, we will copy contents from Reserve to Unreserve in the internal Profile type. This is documented above (Proposal).

In order to work in 1.19 when a v1beta1 config is used, the plugin have to implement both interfaces. Otherwise framework instantiation would error, and thus scheduler startup. When using v1alpha2, it would succeed. I just added a paragraph for this below.